WordPress/openverse-catalog
# Add new columns to MediaStore and database

## Fixes


Fixes WordPress/openverse#1712 by @obulat, fixes WordPress/openverse#1724 , fixes WordPress/openverse#1717 

## Description


This PR combines handling of TSV columns and database columns by using a common `Column` class for both. `Column` class provides the value validation before writing to TSV, and methods for creating database queries (creation, updating, upserting) specific to column types.
As [was discussed ](https://github.com/WordPress/openverse-catalog/discussions/165), we make TSV files use versions, and insert the data that is available in TSV into the database, changing the column order where necessary. The versions I selected are '000' for 'legacy' image TSV, and '001' for current image and audio TSVs, but I'm very open to suggestions on that 😄 

This PR adds or updates the following data columns:
- `category` becomes a common column for both Audio and Image, and is converted from `StringColumn` to an `ArrayColumn` to match the API image category.
- `filetype` is added to both Audio and Image. The provider API scripts will need to be updated to add filetype.
- `filesize`: instead of automatically setting it to None, now we can set it in `ImageStore` and `AudioStore` `.add_` methods.
- `set_position` added to the Audio table.
Before deploying to production, it will be important to add the missing database columns to the production database.

## Testing Instructions

To test these changes, you will probably need to run `just recreate` to make sure that databases are updated to the new version. 

Run the containers, log in to Airflow Web UI. Un-pause `jamendo`, `tsv_to_postgres_loader` and `recreate_audio_popularity_calculation` dags. Jamendo task runs for a very long time, so it's best to check in the logs that some lines have been written to the tsv file:

<img width="702" alt="Screen Shot 2021-09-10 at 4 02 20 PM" src="https://user-images.githubusercontent.com/15233243/132857394-e13a24f0-45a8-4526-99fc-e4ca0db398ca.png">

 and mark the `pull_image_data` as success:
<img width="1023" alt="Screen Shot 2021-09-10 at 3 59 08 PM" src="https://user-images.githubusercontent.com/15233243/132857521-629847a0-e7dd-44dd-8bf8-e38e266f529d.png">

A couple of minutes later, the `tsv_to_postgres_loader` will load the tsv to postgres db. Then you can trigger the `recreate_audio_popularity_calculation` dag to create the audio_view which is the data view that is used by the API.

[Instructions copied from WordPress/openverse-catalog#169]

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
